### PR TITLE
[dev-launcher][iOS] Improve handling of LAN permission crash

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Closes ENG-4203.

# How

Try to retry if the network connection was rejected due to the lack of lan permission.
My code will only retry once per app process - in my opinion, we shouldn't save that information in the shared preferences. Retrying one network call isn't that painful even if the failure wasn't connected with the lan permission. That solution was the best, in my opinion, when I was playing with different versions. 

Also, sometimes when the network call was rejected due to the lack of lan permission, retrying it doesn't help, cause the prompt didn't appear in time, but we can't detect that. However, in that case, the app won't longer crash and also the user will be delegated to the error screen. After all, it seems like pretty good UX. 

# Test Plan

- run bare-expo locally ✅

### CODE REVIEW

**Review by ajsmth — APPROVED** [review]
this looks good - were you able to reproduce and fix the crash with this change?

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m [hunk 1]
```diff
@@ -360,8 +345,24 @@ - (void)loadApp:(NSURL *)expoUrl onSuccess:(void (^ _Nullable)(void))onSuccess o
...
       }
-    } error:onError];
-  } onError:onError];
+    } error:^(NSError * _Nonnull error) {
+      if (@available(iOS 14, *)) {
+        // Try to retry if the network connection was rejected because of the luck of the lan network permission.
+        static BOOL shouldRetry = true;
+        NSString *host = expoUrl.host;
+
+        if (shouldRetry && ([host hasPrefix:@"192.168."] || [host hasPrefix:@"172."] || [host hasPrefix:@"10."])) {
```
Comment (esamelson) [inline packages/expo-dev-launcher/ios/EXDevLauncherController.m:354]: should we add `localhost` to this list, or are simulators not affected by this permission issue?

**Review by esamelson — COMMENTED** [review]
Cool, I agree with your approach and think this looks good -- just a couple thoughts:
- do you think it would be worth adding a short timeout before retrying? This could help catch the cases where the alert doesn't appear in time.
- if not, could we add a short message to the error screen in this case that says something like "You may need to retry if you were just prompted by iOS to allow local network requests."?

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m (hunk 1 above)
Comment (lukmccall) [reply packages/expo-dev-launcher/ios/EXDevLauncherController.m:354]: simulators are not affected by this permission issue ;) So we don't have to do it.

**Comment by lukmccall** [comment]
> this looks good - were you able to reproduce and fix the crash with this change?

Yep, the app won't crash.

**Comment by lukmccall** [comment]
I'll merge that now. 
Also, I'll add the text to the UI in a separate PR.

**Pull request merged**